tests/test_dist_save_load.py

```diff
@@ -73,20 +89,141 @@ def test_distributed_array_save_load(self):
             physical_mesh, load_sharding_spec)
 
         # Check the DistributedArray's remote buffers
-        remote_buffers2 = dist_load_data1.device_mesh.get_remote_buffers(
-            dist_load_data1.remote_buffers, batching=True)
-        desired_buffers2 = [[[0, 1], [2, 3]], [[0, 1], [2, 3]], [[4, 5], [6,
-                                                                          7]],
-                            [[4, 5], [6, 7]]]
-        assert_allclose(np.array(desired_buffers2), np.array(remote_buffers2))
+        desired_buffers2 = np.array([[[0, 1], [2, 3]], [[0, 1], [2, 3]],
+                                     [[4, 5], [6, 7]], [[4, 5], [6, 7]]])
+        self.check_dist_array_eq(desired_buffers2, dist_load_data1)
 
         # Cleanup
         physical_mesh.shutdown()
 
+    def test_distributed_mlp_save_load(self):
+        # Set pipeline parallel option
+        virtual_mesh = DeviceCluster().get_virtual_physical_mesh()
+        set_parallelize_options(devices=virtual_mesh,
+                                strategy="pipeshard_parallel")
+
+        # Init model and optimizer
+        batch_size = 64
+        hidden_dim = 16
+        input_dim = output_dim = hidden_dim
+        model = MLPModel(hidden_dim=hidden_dim,
+                         output_dim=output_dim,
+                         manual_pipeline_layer=True)
+
+        # Init batch args
+        rngkey = jax.random.PRNGKey(0)
+        x = jax.random.normal(rngkey, (batch_size, input_dim))
+        y = jax.random.normal(rngkey, (batch_size, output_dim))
+        batch = {'x': x, 'y': y}
+        state = create_train_state(rngkey, model, [x])
+
+        # Compile
+        global_config.num_micro_batches = 2
+        serial_train_step = get_mlp_train_step(False, None, None, False)
+        parallel_train_step = get_mlp_train_step(True, True, False, False)
+        executable = parallel_train_step.get_executable(state, batch)
+
+        # Run before save
+        serial_state = state
+        parallel_state = state
+        serial_state = serial_train_step(serial_state, batch)[0]
+        parallel_state = parallel_train_step(parallel_state, batch)[0]
+        assert_allclose(serial_state.params, parallel_state.params, 1e-3, 1e-3)
+
+        with tempfile.TemporaryDirectory(
+                prefix="/home/ubuntu/efs/") as ckpt_dir:
```

Comment:
We should not include any absolute path. Can we somehow use ray to hack a temporary NFS just for testing purposes?

Reply:
Or you can force all the testing processes on a same node and just use the filesystem on the node.